io->bio

Koeng101 opened the pull request:
This is a fairly large PR to standardize our parsing and writing of files, using generics to implement higher level functions on simplified interfaces that are standardized across all our readers/writers: genbank, gff, fasta, fastq, slow5, sam, uniprot, rebase, pileup

Koeng101 commented:
This is a work-in-progress, btw

carreter commented:
Why the `io` to `bio` rename?

Koeng101 commented:
> Why the `io` to `bio` rename?

So that the name doesn't conflict with standard library io

TimothyStiles commented on bio/fasta/example_test.go:
```diff
@@ -50,7 +54,7 @@ func ExampleBuild() {
 func ExampleWrite() {
 	fastas, _ := fasta.Read("data/base.fasta")       // get example data
-	_ = fasta.Write(fastas, "data/test.fasta")       // write it out again
+	_ = fasta.WriteFile(fastas, "data/test.fasta")   // write it out again
```
should be just `write`

TimothyStiles commented on bio/bio.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+/*
+This package is supposed to be empty and only exists to provide a doc string.
```
no longer true

TimothyStiles commented on bio/bio.go:
```diff
@@ -0,0 +1,23 @@
+Otherwise its namespace would collide with Go's native IO package.
+*/
+
+type Parser[T any, TH any] interface {
```
What are `T` and `TH` here? Some more explanation on the need for an interface and how it works would be very helpful.

Koeng101 replied:
Types, as are used in generics. I've added in `DataTypes` and `DataTypeHeaders` to be a little more descriptive. While `T` is fairly idiomatic go, `TH` is kinda pushing it, so I think longer descriptions are nice.

TimothyStiles commented on bio/example_test.go:
```diff
@@ -22,7 +22,7 @@ func Example() {
 	_ = gff.Write(gffInput, "test.gff")
 	_ = genbank.Write(gbkInput, "test.gbk")
-	_ = fasta.Write(fastaInput, "test.fasta")
+	_ = fasta.WriteFile(fastaInput, "test.fasta")
```
why `WriteFile` instead of `Write`?

Koeng101 replied:
`Write` in Golang is typically not used on paths (as far as I've seen), but on `io.Writer` interfaces. This change to `WriteFile` when using paths makes it more idiomatic. 

`Write` generally is also being moved into the bio package where we can more effectively test across all parser packages.

TimothyStiles commented on bio/fasta/fasta.go:
```diff
@@ -56,12 +55,6 @@ Keoni
 
 ******************************************************************************/
 
-var (
```
these were used for testing.

Koeng101 replied:
All of these were used in tests that are moving out of `fasta` and into `bio`, where those tests can test all packages at once using generic functions.

carreter commented on bio/fastq/fastq.go:
```diff
@@ -177,6 +170,13 @@ func (parser *Parser) ParseNext() (Fastq, int64, error) {
+	var newSequence []byte
+	for _, char := range sequence {
+		newSequence = append(newSequence, char)
+		if !strings.Contains("ATGCN", string(char)) {
```
```suggestion
		if !strings.ContainsRune("ATGCN", rune(char)) {
```

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,176 @@
+	return nil, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) Parse() ([]DataType, error) {
```
`ParseAll` includes the header. For a lot of formats, the header doesn't really matter (fasta, fastq, etc), for some it matters a lot (slow5, etc)

carreter replied:
Ah, gotcha. Maybe call it `ParseWithHeader` instead?

Koeng101 replied:
Makes sense, changing

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,176 @@
+	Next() (DataType, int64, error)
+}
+
+type Writer interface {
```
Ty, didn't look there, changed the type signatures.

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,176 @@
+
+/******************************************************************************
+
+Writer functions
```
`io.MultiWriter()` has a type signature of `MultiWriter(writers ...Writer) Writer`, whereas `WriteAll` has a type signature of `WriteAll(data []io.WriterTo, header io.WriterTo, w io.Writer) error`. The writerTo are different, and generally they do different things: `WriteAll` writes many records + header to a single file, while `io.MultiWriter` splits one write to many writers, kinda like `tee`.

Koeng101 replied:
`WriteFile()` is just legacy - pretty much all our previous packages have it. @TimothyStiles thoughts on including a `WriteFile()`? I'm leaning towards removing it, since os.Open is pretty simple.

Koeng101 commented on bio/bio.go:
```diff
@@ -4,20 +4,173 @@ Package bio provides utilities for reading and writing sequence data.
+// longer. We use the default maxLineLength from bufio unless there is a
+// particular reason to use a different number.
+const defaultMaxLineLength int = bufio.MaxScanTokenSize // 64kB is a magic number often used by the Go stdlib for parsing.
+var (
```
Implemented, good catch.

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,176 @@
+	return p.LowLevelParser.Header()
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseN(countN int) ([]DataType, error) {
```
Implemented all. Mostly it was just copy pasting the copy-pastes around from other parsers...

Koeng101 commented on bio/bio.go:
```diff
@@ -4,20 +4,173 @@ Package bio provides utilities for reading and writing sequence data.
+
+******************************************************************************/
+
+type LowLevelParser[DataType fasta.Fasta, DataTypeHeader fasta.Header] interface {
```
Hmm I agree we should probably change `LowLevelParser`, but Concrete isn't quite there for me. My first reaction was "Concrete, like a foundation, like the lowest level?"

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,176 @@
+
+******************************************************************************/
+
+func (p *Parser[DataType, DataTypeHeader]) Next() (DataType, int64, error) {
```
I had the same thoughts actually, I'm like 50/50 on `int64`. The reason I added that was to stay consistent with `io.Read`, and for the fact that you could use the numbers as a seek for larger files (like save the offsets)

carreter replied:
We are already not consistent with `io.Read`. The `int64` is only really necessary to know how many bytes were read into the buffer you pass to `io.Read`.

Koeng101 replied:
> The int64 is only really necessary to know how many bytes were read into the buffer you pass to io.Read

True, I'm convinced. Changing.

Koeng101 replied:
removed

Koeng101 commented:
`fasta` parser testing is now at 98.5%, other than the Scanner non-EOF error, which I can't really figure out how to test after a very brief trying-to-figure-out. I removed a ton of code, and everything should be fairly simple now.

carreter commented on bio/bio.go:
```diff
@@ -0,0 +1,195 @@
+
+func NewParser(format Format, r io.Reader) (*Parser[fasta.Record, fasta.Header], error) {
+	var maxLineLength int
+	switch format {
```
Since we have a map of max lengths now, this switch isn't necessary.

Koeng101 replied:
True, fixing!

carreter commented on bio/bio.go:
```diff
@@ -0,0 +1,195 @@
+	return data, header, nil
+}
+
+func (p *Parser[DataType, DataTypeHeader]) ParseConcurrent(channel chan<- DataType) error {
```
As written, this non-current and blocking. You probably want to wrap the for loop in a `go func(){ ... }()` to make it nonblocking and call it `ParseToChannel()` instead. Make sure to note whether or not it is block in that case.

Koeng101 replied:
Concurrency in term of parsing / IO is kind of a lie though, because you are literally reading the file with ` io.Reader`. I do agree though, bad name, I like the name `ParseToChannel`

(removed the `close(channel)` for below stuff)

The change to `go func(){}()` would make it nonblocking, but also make error checking more difficult. I'm thinking of using this with something like [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup), so that you can begin parsing multiple files while still having error handling in a pretty simple-to-understand way.

In the case that you are parsing one file, IO will likely be your limiting factor, and a blocking function is pretty easy to use and understand how it works. As you need to scale, you can just wrap the overall function in `go func(){}()`.


I am open to making it non-blocking tho - is there any particular reason you have in mind for making it non-blocking?

carreter commented on bio/fasta/fasta.go:
```diff
@@ -51,84 +50,47 @@ Keoni
 	// reader keeps state of current reader.
-	reader bufio.Reader
-	line   uint
+	reader     bufio.Scanner
```
Call this `scanner` instead

Koeng101 commented on bio/fastq/fastq.go:
```diff
@@ -0,0 +1,212 @@
+	line   uint
+}
+
+// Header returns a identifier with nothing and nil.
```
remove 'a identifier'

Koeng101 commented:
fixes #325

Koeng101 commented:
fixes #327

Koeng101 commented:
I'm gonna say this is probably no longer a work-in-progress. @TimothyStiles review?

carreter commented on bio/fasta/fasta_test.go:
```diff
@@ -0,0 +1,180 @@
+func TestHeader(t *testing.T) {
+	_, err := NewParser(nil, 256).Header()
+	if err != nil {
+		t.Errorf("Header should never be nil")
```
```suggestion
		t.Errorf("Unexpected error while parsing header: %v", err)
```

carreter reviewed (commented):
Can we move all the `bio/<format>/data` files to `test/data` or `testdata/`?

carreter commented on bio/bio.go:
```diff
@@ -0,0 +1,307 @@
+// pointer is used to represent the difference between a null DataType and an
+// empty DataType.
+type LowLevelParser[DataType fasta.Record | fastq.Read | slow5.Read | pileup.Line, DataTypeHeader fasta.Header | fastq.Header | slow5.Header | pileup.Header] interface {
+	Header() (*DataTypeHeader, error)
```
Why does `LowLevelParser` return pointers and `Parser` return values?

Koeng101 replied:
I think both return pointers. Could you link where `Parser` returns values?

carreter replied:
Oh my bad I was misreading. Marking resolved.

carreter replied:
Actually, why are we returning pointers to the structs instead of the structs themselves?

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,307 @@
+	Next() (*DataType, error)
+}
+
+// The following checks that all DataType and DataTypeHeaders implement the io.WriteTo interface.
```
Can you link that comment?

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,307 @@
+// for this is needed at the last Next(), when it returns an io.EOF error. A
+// pointer is used to represent the difference between a null DataType and an
+// empty DataType.
+type LowLevelParser[DataType fasta.Record | fastq.Read | slow5.Read | pileup.Line, DataTypeHeader fasta.Header | fastq.Header | slow5.Header | pileup.Header] interface {
```
I renamed to `parserInterface`, because it is the interface that the Parsers have to fulfill

Koeng101 commented:
> Can we move all the `bio/<format>/data` files to `test/data` or `testdata/`?

The difficulty there would be if any of the specific parsers require the testdata, they can't really import `..`. So we'd have to have 2 copies, or we just have different test data for the specific parsers and generic parsers.

Koeng101 commented:
> > Can we move all the `bio/<format>/data` files to `test/data` or `testdata/`?
> 
> The difficulty there would be if any of the specific parsers require the testdata, they can't really import `..`. So we'd have to have 2 copies, or we just have different test data for the specific parsers and generic parsers.

I'm completely wrong here actually, I was thinking of go:embed. Yes we can do this

Koeng101 commented:
```
Auto-merging bio/genbank/example_test.go
Auto-merging bio/genbank/genbank_test.go
Auto-merging bio/gff/example_test.go
Auto-merging bio/pileup/pileup.go
CONFLICT (content): Merge conflict in bio/pileup/pileup.go
Auto-merging bio/pileup/pileup_test.go
CONFLICT (content): Merge conflict in bio/pileup/pileup_test.go
Auto-merging bio/polyjson/example_test.go
Auto-merging bio/slow5/slow5.go
Auto-merging bio/uniprot/example_test.go
CONFLICT (content): Merge conflict in bio/uniprot/example_test.go
CONFLICT (modify/delete): io/example_test.go deleted in HEAD and modified in main.  Version main of io/example_test.go left in tree.
CONFLICT (modify/delete): io/fasta/fasta.go deleted in HEAD and modified in main.  Version main of io/fasta/fasta.go left in tree.
Auto-merging seqhash/seqhash_test.go
Auto-merging synthesis/codon/codon_test.go
Auto-merging synthesis/codon/example_test.go
Automatic merge failed; fix conflicts and then commit the result.
```

Good lord, going through and fixing.

TimothyStiles commented on bio/example_test.go:
```diff
@@ -0,0 +1,234 @@
+	// Poly can also output these file formats. Every file format has a corresponding Write function.
+	_ = gff.Write(gffInput, "test.gff")
+	_ = genbank.Write(gbkInput, "test.gbk")
+	//_ = fasta.WriteFile(fastaInput, "test.fasta")
```
why is fasta commented out?

Koeng101 replied:
No `Write` functions anymore (see: `Example_writeAll()`)

Koeng101 replied:
We *could* reimplement them, but the type checking will kinda suck, or we'd have to be super verbose. The reason is that `[]*DataType` does not equal `[]interface{}`, even if `*DataType` DOES equal `interface{}`. Because its a slice, you have to either accept `any` as an input, or you have to do an interface conversion beforehand, which is kinda the worst.

Alternatively, they all implement `io.WriterTo`, so you can just write it to a document using a standard-library  interface.

TimothyStiles commented on bio/example_test.go:
```diff
@@ -0,0 +1,234 @@
+
+	gffInput, _ := gff.Read("../data/ecoli-mg1655-short.gff")
+	gbkInput, _ := genbank.Read("../data/puc19.gbk")
+	//fastaInput, _ := fasta.Read("fasta/data/base.fasta")
```
why is fasta commented out?

Koeng101 replied:
No `Read` functions anymore.

Koeng101 replied:
Those could be added back in for each and every parser, it's just a bit annoying because how generic initialization works (ie, there would need to be a function for every type, like how parsing is done)

TimothyStiles commented on bio/example_test.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func Example_readGz() {
+	// ReadGz lets you read gzipped files into a parser.
```
ReadGz isn't a function anymore. Update comment.

TimothyStiles commented on bio/example_test.go:
```diff
@@ -0,0 +1,234 @@
+	// First, lets make a file that is gzip'd, represented by this
+	// buffer.
+	var file bytes.Buffer
+	zw := gzip.NewWriter(&file)
```
`zw` stands for "zip writer". Unclear.

TimothyStiles commented on bio/example_test.go:
```diff
@@ -0,0 +1,234 @@
+	// Output: ADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTIDFPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREADIDGDGQVNYEEFVQMMTAK*
+}
+
+func ExampleParser_ParseWithHeader() {
```
Am liking the interface overhaul overall 👍

TimothyStiles commented on bio/pileup/pileup.go:
```diff
@@ -60,12 +56,12 @@ type Pileup struct {
+// Header is a blank struct, needed for compatibility with bio parsers. It contains nothing.
+type Header struct{}
+
+// WriteTo is a blank function, needed for compatibility with bio parsers. It doesn't do anything.
```
Why no writer?

Koeng101 replied:
There is a writer, `WriteTo` (needed for interface compatibility). Pileup files don't have headers though, so in this case it doesn't do anything.

TimothyStiles commented on bio/pileup/pileup.go:
```diff
@@ -74,47 +70,25 @@ type Parser struct {
 	line   uint
 }
 
+// Header returns a identifier with nothing and nil.
```
?

TimothyStiles commented on bio/bio.go:
```diff
@@ -0,0 +1,307 @@
+	"github.com/TimothyStiles/poly/bio/fasta"
+	"github.com/TimothyStiles/poly/bio/fastq"
+	"github.com/TimothyStiles/poly/bio/pileup"
+	"github.com/TimothyStiles/poly/bio/slow5"
```
No genbank or gff?

TimothyStiles reviewed (commented):
Not sure why gff and genbank are not included in `bio.go`.

Koeng101 commented:
@TimothyStiles Just made the genbank parser compatible. All it cost was... nuking all of the tests, because they pretty much all relied on read functions. Those I'll have to add back in later.

Koeng101 commented:
Ah crap, other files depend on `genbank.Read`... I'll fix later.

carreter commented on bio/bio.go:
```diff
@@ -0,0 +1,323 @@
+	return manyToChannel[slow5.Read, slow5.Header](ctx, parsers, channel)
+}
+
+// ManyPileupToChannel takes a list of pileup parsers and pipes all
+// records/reads/lines to a single channel.
+func ManyPileupToChannel(ctx context.Context, parsers []*Parser[pileup.Line, pileup.Header], channel chan<- pileup.Line) error {
+	return manyToChannel[pileup.Line, pileup.Header](ctx, parsers, channel)
+}
```
`ParseToChannel` already checks for `ctx.Done()` so this is unnecessary.

Koeng101 replied:
True. Deleted.

carreter commented on bio/bio.go:
```diff
@@ -0,0 +1,323 @@
+		default:
+			record, err := p.Next()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					err = nil // EOF not treated as parsing error.
+				}
+				if record != nil {
+					channel <- *record
+				}
+				if !keepChannelOpen {
+					close(channel)
+				}
+				return err
+			}
+			channel <- *record
+		}
+	}
+}
+
+// ManyFastaToChannel takes a list of FASTA parsers and pipes all
+// records/reads/lines to a single channel.
+func ManyFastaToChannel(ctx context.Context, parsers []*Parser[fasta.Record, fasta.Header], channel chan<- fasta.Record) error {
+	return manyToChannel[fasta.Record, fasta.Header](ctx, parsers, channel)
+}
+
+// ManyFastqToChannel takes a list of FASTQ parsers and pipes all
```
This functions are unnecessary as `manyToChannel` is already generic.

Koeng101 replied:
You'd have to do a generic initialization on `manyToChannel` though, and I'm not sure we should have our users have to figure out generics

carreter replied:
Is it necessary? I think the compiler might be able to infer the types, at least my IDE was indicating that calling `manyToChannel` didn't require the generic initialization.

Koeng101 replied:
Ya know what, I guess I haven't tried, since it could infer the type from the channel input. Lemme test it out, I might be WRONG and DUMB

Koeng101 replied:
u right, you don't have to do generic initialization and I am just bad at generics

carreter reviewed (changes requested):
I haven't reviewed the individual formats yet, will come back and do that later.

carreter commented:
Crap, seems like the line numbers on my last review got messed up because I hadn't synced VSCode with the remote repo. Let me know if you have any questions @Koeng101

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,323 @@
+
+Higher level parse
+
+******************************************************************************/
```
This is a really really good point and I can't believe I didn't think of this, makes a lot of sense. Changing! (gotta go change all the NewParser funcs tho)

Koeng101 replied:
So the problem is:

The parsers assume that you can return a `Data, io,EOF` and for that data to still be valid. This mainly occurs when an EOF occurs before a newline at the end of some data, which is fairly common in the test cases. In order to use an `io.WriterTo`, we would have to change all the underlying parsers to only return valid `Data` OR `io.EOF`, not AND. 

This is doable, but will take me changing the 5 or so underlying parsers. Overall, it makes sense to me, so I'll do this, but it'll take a little.

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,323 @@
+// close channel once all are done parsing.
+//
+// Context can be used to close the parser in the middle of parsing - for
+// example, if an error is found in another parser elsewhere and all files
+// need to close.
+func (p *Parser[DataType, DataTypeHeader]) ParseToChannel(ctx context.Context, channel chan<- DataType, keepChannelOpen bool) error {
```
Well, originally, it was to simplify the use of the channel. Ie, you should be able to run `go ParseToChannel` and then when the parsing is done it'll clean up the channel for you. Otherwise, your for loop wouldn't really end when you start receiving from the channel, and you have no way to know if it terminated, unless you put in your own waitgroup or whatever

Koeng101 commented:
- Converted all types to io.Writer in parser. All the parsers have to return is something that fulfills io.Writer for them to be valid. 
- Changed return types on all parsers to NOT ever return an io.EOF and valid `Data` at the same time.
- Genbank merged, all other dependencies fixed.

- [ ] Add back Genbank tests
- [ ] Merge gff
- [ ] Merge uniprot
- [ ] Add comprehensive EOF/newline tests for all parsers.
- [ ] Update comments about `Next()` function

carreter commented:
Don't forget to credit @folago :

Co-authored-by: name <name@example.com>

carreter commented:
Fixes #364 .

Koeng101 commented:
Ok, it is more important I think, to get this merged than be perfect. I'm doing the following.

- Removing the gff parser, for now. We can add it back later.
- I have not merged uniprot. We'll have to do that later as we standardize everything
- Add comprehensive EOF/newline tests for all parsers will have to be done later

I will be merging genbank tests and Next comments. Upcoming...

Koeng101 commented:
If we want to go ahead with this, we would spawn 3 new issues:

- Make compatible uniprot parser
- Make compatible gff parser
- Add comprehensive EOF / newline parser tests

abondrn reviewed (commented):
Overall looks great, I only had stylistic comments. Due to the size of the PR and it being a pareto improvement (common interface, strictly greater testing, no loss in efficiency) I'm leaning towards approving ASAP. Since all of the key contributors have had their say, for the remaining change requests simply determine whether it is (1) a quick fix (2) issue worthy (3) won't do. No additional features please.

Koeng101 commented on seqhash/seqhash_test.go:
```diff
@@ -66,7 +67,10 @@ func TestHash(t *testing.T) {
 func TestLeastRotation(t *testing.T) {
-	sequence, _ := genbank.Read("../data/puc19.gbk")
+	file, _ := os.Open("../data/puc19.gbk")
+	defer file.Close()
```
I thought that was how it worked, but code I've seen still usually has the defer Close! Do you know why?

abondrn replied:
Started thread in Discord

Koeng101 commented on io/io.go:
Gotcha, will do in the future. io/io.go has basically entirely died, so I'm fine with that here.

Usually git is able to figure out if I moved a file if I just immediately commit again

Koeng101 commented on go.mod:
```diff
@@ -15,8 +15,11 @@ require (
 
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/intel-go/cpuid v0.0.0-20181003105527-1a4a6f06a1c6 // indirect
```
I think it's just used in `StreamVByte`, but I don't think any other file format actually would benefit from this kind of compression very much. It is very nice in slow5 though, because of the integer data streaming out of the nanopore.

Koeng101 commented on bio/bio.go:
```diff
@@ -0,0 +1,275 @@
+/*
```
Let's definitely make an issue for that!

Koeng101 commented on bio/genbank/genbank.go:
```diff
@@ -1026,3 +1086,86 @@ func generateWhiteSpace(length int) string {
+
+/******************************************************************************
+
+Old functions for testing here.
```
> If we are only keeping these for backwards test compatibility, why not just keep the capitalization?

Simplify the user API (for GoDocs and the such). We don't want people actually using those functions, and I'd prefer we get rid of them - it was just too much to do right now.

Koeng101 commented on bio/genbank/genbank.go:
```diff
@@ -33,7 +33,7 @@ GBK specific IO related things begin here.
 
 var (
 	readFileFn        = os.ReadFile
```
I guess it is a pattern, but I'm not a huge fan either. It's something I've just pulled along from older tests

Koeng101 commented on bio/genbank/genbank.go:
```diff
@@ -412,136 +449,159 @@ func (params *parseLoopParameters) init() {
+// Parser is a genbank parser created on an io.Reader.
+type Parser struct {
+	scanner    bufio.Scanner
+	parameters parseLoopParameters
```
Good point. The reason is basically historical, so I should fix that.

Koeng101 commented on bio/fasta/fasta.go:
```diff
@@ -0,0 +1,219 @@
+Fasta is a very simple file format for working with DNA, RNA, or protein sequences.
+It was first released in 1985 and is still widely used in bioinformatics.
+
+https://en.wikipedia.org/wiki/_format
```
Definitely, need to fix that

Koeng101 commented on bio/fasta/fasta.go:
```diff
@@ -0,0 +1,219 @@
+
+https://en.wikipedia.org/wiki/_format
+
+One interesting use of the concurrent  parser is working with the Uniprot
```
thank you for pointing that out

Koeng101 commented:
TODO:
- Internalize parse parameters
- Add back fasta format in docs
- Remove that extra space

Koeng101 commented:
> TODO:
> 
>     * Internalize parse parameters
> 
>     * Add back fasta format in docs
> 
>     * Remove that extra space

Parse parameters are actually there for reason. it's reset fairly often, so it is easier to have it outside the parser itself. We could have a separate method for fully resetting it, but I think it is fairly readable as is.

Koeng101 commented:
Alright, looking over this it looks good to get a review from @TimothyStiles

TimothyStiles commented:
@Koeng101

I like the idea of utilizing generics but am not too familiar with their usage and syntax. Can you add a little "how-to" guide for writing parsers and writer in this new setup?

Koeng101 commented:
> @Koeng101
> 
> I like the idea of utilizing generics but am not too familiar with their usage and syntax. Can you add a little "how-to" guide for writing parsers and writer in this new setup?

Hmmm, for both of those it doesn't really have much to do with generics (mostly just interfaces) - you only need generics once you make the top level convenience-function, which is essentially copy-pasted over each parser type, because Go generics are limited. Really what you have to be concerned about when writing parsers is the interfaces they must implement. To be clearer with the generics:

```go
// NewFastaParser initiates a new FASTA parser from an io.Reader.
func NewFastaParser(r io.Reader) (*Parser[*fasta.Record, *fasta.Header], error) {
	return NewFastaParserWithMaxLineLength(r, DefaultMaxLengths[Fasta])
}

// NewFastaParserWithMaxLineLength initiates a new FASTA parser from an
// io.Reader and a user-given maxLineLength.
func NewFastaParserWithMaxLineLength(r io.Reader, maxLineLength int) (*Parser[*fasta.Record, *fasta.Header], error) {
	return &Parser[*fasta.Record, *fasta.Header]{parserInterface: fasta.NewParser(r, maxLineLength)}, nil
}

// NewFastqParser initiates a new FASTQ parser from an io.Reader.
func NewFastqParser(r io.Reader) (*Parser[*fastq.Read, *fastq.Header], error) {
	return NewFastqParserWithMaxLineLength(r, DefaultMaxLengths[Fastq])
}

// NewFastqParserWithMaxLineLength initiates a new FASTQ parser from an
// io.Reader and a user-given maxLineLength.
func NewFastqParserWithMaxLineLength(r io.Reader, maxLineLength int) (*Parser[*fastq.Read, *fastq.Header], error) {
	return &Parser[*fastq.Read, *fastq.Header]{parserInterface: fastq.NewParser(r, maxLineLength)}, nil
}
```

What you can notice from both of these is that the Fastq is literally the same as the Fasta, except the `Data` and `Header` now points to fastq types rather than fasta types. The only reason you have to copy and paste this all the time instead of giving the format as a variable into a function is due to the limitation of Golang types.

## writers + parsers
"How to" make a parser is almost entirely contained within these few lines (as part of the type system):
```go
type parserInterface[Data io.WriterTo, Header io.WriterTo] interface {
	Header() (Header, error)
	Next() (Data, error)
}

type Parser[Data io.WriterTo, Header io.WriterTo] struct {
	parserInterface parserInterface[Data, Header]
}
```

Let's dive into the Golang type system to understand it better. A Parser is a generic struct, which means it is *any* struct that implements a `parserInterface`. What does that mean? Well, it means that any struct that implements the methods `Next()` and `Header()` satisfies the `parserInterface`. Both `Next()` and `Header()` just return an `io.WriterTo` and err, which is a standard library interface + primitive.

To sum it up, a `Parser` is *any* struct that returns a `io.WriterTo`+err when you call `Header()` or `Next()`. It is extremely simple. 

For example, the following parser would satisfy the `bio.Parser`:
```go
type SimpleHeader struct {}
type SimpleData struct {}
func (h SimpleHeader) WriteTo(w io.Writer) (n int64, err error) { return 0, nil}
func (d SimpleData) WriteTo(w io.Writer) (n int64, err error) { return 0, nil}

type SimpleParser struct{}
func (p SimpleParser) Header() (SimpleHeader, error) { return SimpleHeader{}, nil}
func (p SimpleParser) Next() (SimpleData, error) { return SimpleData{}, nil }
```

In this example, we have all the fundamentals needed:
1. A Header
2. A Data
3. An `io.WriteTo` implemented on the Header
4. An `io.WriteTo` implemented on the Data
5. A Parser
6. A `Header()` implemented on the parser
7. A `Next()` implemented on the parser

The clever bit is realizing that pretty much every single biological data structure has something along these lines, and that `Next()` supports stream-reading - and as a side effect of implementing the `io.WriteTo` interface, you also have writing built-in to the type definitions in a standard-library supported way.

In this way, the interfaces here are essentially just forcing parsers to expose 4 functions (`WriteTo` x2, `Next`, `Header`), and by doing that on their underlying data structures, you now can have a completely standardized way of working with each parser. 

## wrapping it up

Let's say you want to implement `SimpleParser` as a top level parser. Here is how you would do that:

```go
func NewSimpleParser() (*Parser[*SimpleData, *SimpleHeader], error) {
   return &Parser[*SimpleData, *SimpleHeader]{parserInterface: SimpleParser{}}, nil
}
```

All this does is:
1. Say that we are returning a Parser with the two types `SimpleData` and `SimpleHeader`
2. Implements the `parserInterface` needed for a Parser (because `SimpleParser` has those handy `Next()` and `Header()` functions)
3. returns the instantiated generic Parser, wrapping that underlying `SimpleParser`

## why we couldn't do this before generics

The hard bit is that I wanted to expose the underlying data types when you do things with a Parser. To pull from some above examples:

```go
parser, _ := bio.NewFastaParser(file)
fastaRecord, _ := parser.Next()
// you should be able to do the below!
fmt.Println(fastaRecord.Sequence)
fmt.Println(fastaRecord.Read)
```

Without a generic interface, we couldn't say that it was aight for `Data` to contain other things, like `Sequence` or `Name`, beyond the required `WriteTo`. Essentially, all the generics are doing is saying it is aight for `Data` to contain other things without having to go through complicated type conversion magic.

paleale commented:
Hello! Can I ask you to see and apply my changes for #352 ? 
I don't know a proper way for such changes to be suggested, so I've prepared a patch:
[0001-remove-not-working-error-handling-from-GetSequence-f.patch](https://github.com/TimothyStiles/poly/files/13464941/0001-remove-not-working-error-handling-from-GetSequence-f.patch)
Changes in CONTRIBUTION.md were discussed in #352 also... Thank you!

Koeng101 commented:
> Hello! Can I ask you to see and apply my changes for #352 ? I don't know a proper way for such changes to be suggested, so I've prepared a patch: [0001-remove-not-working-error-handling-from-GetSequence-f.patch](https://github.com/TimothyStiles/poly/files/13464941/0001-remove-not-working-error-handling-from-GetSequence-f.patch) Changes in CONTRIBUTION.md were discussed in #352 also... Thank you!

Sure! Could you add a pull request with your changes into this branch? That would be the proper way + our tooling for reviewing changes would work well with that. Normally, you just make the changes on your own fork, and then do a pull request into this branch.

isaacguerreir commented:
It would be interesting now that we're using generics to have a factory to decide which parser to use based on the file content, so the developer doesn't need to choose what parser to use.

Koeng101 commented:
> It would be interesting now that we're using generics to have a factory to decide which parser to use based on the file content, so the developer doesn't need to choose what parser to use.

Due to how generics work in Go, I'm not sure that you actually can do this. Basically, if the output signature is `Parser[Data, Header]`, you can't return `Parser[fasta.Name, fasta.Header]`.

isaacguerreir commented:
my mistake, for some reason reading your explanation I thought it would be possible to pass generic types as outputs, but taking a better look at the documentation it seems undoable to use Golang generics like that. It would be possible if the function returns a common structure, which is not possible with the current implementation.

Koeng101 commented:
> seems undoable to use Golang generics like that. It would be possible if the function returns a common structure, which

Yes, we don't have a common structure between all of the different parsers, and I think this is a good thing. Makes each data type very specific and concise.
